server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,438 @@
+            public void onFailure(String source, Exception e) {
+                logger.error(
+                    () -> new ParameterizedMessage(
+                        "failed to start decommission action for attribute [{}]",
+                        decommissionAttribute.toString()
+                    ),
+                    e
+                );
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                DecommissionAttributeMetadata decommissionAttributeMetadata = newState.metadata()
+                    .custom(DecommissionAttributeMetadata.TYPE);
+                assert decommissionAttribute.equals(decommissionAttributeMetadata.decommissionAttribute());
+                decommissionClusterManagerNodes(decommissionAttributeMetadata.decommissionAttribute(), listener);
+            }
+        });
+    }
+
+    private void decommissionClusterManagerNodes(
+        final DecommissionAttribute decommissionAttribute,
+        ActionListener<ClusterStateUpdateResponse> listener
+    ) {
+        ClusterState state = clusterService.getClusterApplierService().state();
+        Set<DiscoveryNode> clusterManagerNodesToBeDecommissioned = filterNodesWithDecommissionAttribute(state, decommissionAttribute, true);
+        try {
+            // this is a sanity check that the cluster will not go into a quorum loss state because of exclusion
+            ensureNoQuorumLossDueToDecommissioning(
+                decommissionAttribute,
+                clusterManagerNodesToBeDecommissioned,
+                state.getLastCommittedConfiguration()
+            );
+        } catch (DecommissioningFailedException dfe) {
+            listener.onFailure(dfe);
+            decommissionController.updateMetadataWithDecommissionStatus(
+                DecommissionStatus.FAILED,
+                new ActionListener<DecommissionStatus>() {
+                    @Override
+                    public void onResponse(DecommissionStatus status) {
+                        logger.info(
+                            "updated the status to [{}], as cluster could have gone to quorum loss situation due to decommissioning",
+                            status.toString()
+                        );
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.error("unexpected error found while updating the status", e);
+                    }
+                }
+            );
+        }
+        // remove all 'to-be-decommissioned' cluster manager eligible nodes from voting config
+        // The method ensures that we don't exclude same nodes multiple times
+        boolean toBeDecommissionedClusterManagerNodesExcluded = excludeDecommissionedClusterManagerNodesFromVotingConfig(
+            clusterManagerNodesToBeDecommissioned
+        );
+
```

Comment:
Yes, this makes sense. Updating the logic in next iteration to - 
```
private void decommissionClusterManagerNodes(
        final DecommissionAttribute decommissionAttribute,
        ActionListener<ClusterStateUpdateResponse> listener
    ) {
        ClusterState state = clusterService.getClusterApplierService().state();
        Set<DiscoveryNode> clusterManagerNodesToBeDecommissioned = filterNodesWithDecommissionAttribute(state, decommissionAttribute, true);
        try {
            // this is a sanity check that the cluster will not go into a quorum loss state because of exclusion
            ensureNoQuorumLossDueToDecommissioning(
                decommissionAttribute,
                clusterManagerNodesToBeDecommissioned,
                state.getLastCommittedConfiguration()
            );
        } catch (DecommissioningFailedException dfe) {
            listener.onFailure(dfe);
            decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.FAILED, statusUpdateListener());
            return;
        }

        ActionListener<Void> exclusionListener = new ActionListener<Void>() {
            @Override
            public void onResponse(Void unused) {
                if (transportService.getLocalNode().isClusterManagerNode()
                    && !nodeHasDecommissionedAttribute(transportService.getLocalNode(), decommissionAttribute)
                ) {
                    // we are good here to send the response now as the request is processed by an eligible active leader
                    // and to-be-decommissioned cluster manager is no more part of Voting Configuration
                    listener.onResponse(new ClusterStateUpdateResponse(true));
                    failDecommissionedNodes(clusterService.getClusterApplierService().state());
                } else {
                    // explicitly calling listener.onFailure with NotClusterManagerException as we can certainly say that
                    // the local cluster manager node will be abdicated and soon will no longer be cluster manager.
                    // this will ensure that request is retried until cluster manager times out
                    listener.onFailure(
                        new NotClusterManagerException(
                            "node ["
                                + transportService.getLocalNode().toString()
                                + "] not eligible to execute decommission request. Will retry until timeout."
                        )
                    );
                }
            }

            @Override
            public void onFailure(Exception e) {
                decommissionController.updateMetadataWithDecommissionStatus(DecommissionStatus.FAILED, statusUpdateListener());
            }
        };

        // remove all 'to-be-decommissioned' cluster manager eligible nodes from voting config
        Set<String> nodeIdsToBeExcluded = clusterManagerNodesToBeDecommissioned.stream()
            .map(DiscoveryNode::getId)
            .collect(Collectors.toSet());

        final Predicate<ClusterState> allNodesRemoved = clusterState -> {
            final Set<String> votingConfigNodeIds = clusterState.getLastCommittedConfiguration().getNodeIds();
            return nodeIdsToBeExcluded.stream().noneMatch(votingConfigNodeIds::contains);
        };
        if (allNodesRemoved.test(clusterService.getClusterApplierService().state())) {
            exclusionListener.onResponse(null);
        } else {
            // send a transport request to exclude to-be-decommissioned cluster manager eligible nodes from voting config
            decommissionController.excludeDecommissionedNodesFromVotingConfig(
                nodeIdsToBeExcluded,
                new ActionListener<Void>() {
                    @Override
                    public void onResponse(Void unused) {
                        logger.info(
                            "successfully removed decommissioned cluster manager eligible nodes [{}] from voting config ",
                            clusterManagerNodesToBeDecommissioned.toString()
                        );
                        exclusionListener.onResponse(null);
                    }

                    @Override
                    public void onFailure(Exception e) {
                        logger.debug(
                            new ParameterizedMessage("failure in removing decommissioned cluster manager eligible nodes from voting config"),
                            e
                        );
                        exclusionListener.onFailure(e);
                    }
                }
            );
        }
    }
```